### ISSUE TITLE
"Replaced" packages cannot be reverted to its original

### ISSUE THREAD

**Issue by JanSlabon** [composer/composer#7809]
When removing a dependency that replaces another package the dependency to the original package is never restored.

I prepared a complete reproducable test setup. So the repositories in the following composer.json versions will work so you can play with this issue. For completion all used files are available [here](https://github.com/composer/composer/files/2624840/poc.zip).

Let's say I have a dependency to "setasign-composer-test/checker" which depends on "setasign-composer-test/root":
```json
{
    "config": {
        "secure-http": false
    },
    "repositories": [
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo1"
        },
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo2"
        }
    ],
    "require": {
        "setasign-composer-test/checker": "*"
    }
}
```

When I run `composer install` I get the following output:
```
Loading composer repositories with package information
Warning: Accessing customers.setasign.com over http which is an insecure protocol.
Updating dependencies (including require-dev)
Package operations: 2 installs, 0 updates, 0 removals
  - Installing setasign-composer-test/root (v1.0.0): Downloading (100%)
  - Installing setasign-composer-test/checker (v1.0.0): Downloading (100%)
Writing lock file
Generating autoload files
```

Executing a test script with following content:
```php
<?php
require_once 'vendor/autoload.php';

new setasign\composer_bug\Checker();
```

I get the expected output:
```
I'm version Nr #1
```
...which is the output in "setasign-composer-test/root".

Now I want use another package, that replaces "setasign-composer-test/root":
```json
{
    "config": {
        "secure-http": false
    },
    "repositories": [
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo1"
        },
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo2"
        }
    ],
    "require": {
        "setasign-composer-test/replacement": "*",
        "setasign-composer-test/checker": "*"
    }
}
```

After calling `composer update` I get:
```
Loading composer repositories with package information
Warning: Accessing customers.setasign.com over http which is an insecure protocol.
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating setasign-composer-test/replacement (v1.0.0 => v1.0.0): Downloading (100%)
Writing lock file
Generating autoload files
```

When I call the test script now I get:
```
I'm version Nr #2
```

Which is the output form "setasign-composer-test/replacement". Perfect, the package was replaced successfully.

Now comes the faulty part: When I reset the composer.json to the inital version by removing the dependency to the replacement package:

```json
{
    "config": {
        "secure-http": false
    },
    "repositories": [
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo1"
        },
        {
            "type": "composer",
            "url": "http://customers.setasign.com/composer/replace/repo2"
        }
    ],
    "require": {
        "setasign-composer-test/checker": "*"
    }
}
```

and I call `composer update` I get following output:
```
Loading composer repositories with package information
Warning: Accessing customers.setasign.com over http which is an insecure protocol.
Updating dependencies (including require-dev)
Nothing to install or update
Generating autoload files
```

Nothing happens after all. The call of the test script still outputs:
```
I'm version Nr #2
```

When asking composer:
```
composer why "setasign-composer-test/replacement"
```

I get:
```
There is no installed package depending on "setasign-composer-test/replacement"
```

The only way to get rid of this dependency is by removing the origin "setasign-composer-test/checker" from the composer.json file, update, reintroduce the dependency and update again. Or (for sure) by deleting the vendor folder...

I guess this is related to #7573 but minized to a more simple setup.

**Comment by stof** [comment]
this issue is a duplicate of #7573

**Comment by alcohol** [comment]
Yeah I am not sure what to do with this issue. It is a good reproducible scenario, but it is also very much a duplicate of #7573. Do you mind if we close this?

**Comment by JanSlabon** [comment]
I had this problem in a production workflow (replacing encrypted evaluation versions with final versions) and re-build this to ensure that it's not our fault. Before I leave this discoveries under the desk in hope that #7573 will be fixed soon I though I report it with a complete reproducable test case to give it some more attention. If you want to close it in favor of #7573 I'm fine with this :-) I just thought that I could help with a complete independent and minimal test case ;-)

**Comment by stof** [comment]
well, #7573 is fully known. But solving it requires changing the way dependency resolution works in composer, which is why it is part of the 2.0 milestone.

**Comment by Seldaek** [comment]
We can keep this too to make sure the repro case works once 2.0 is more ready.. The 2.0 milestone is a huge mess anyway right now :)

**Comment by Seldaek** [comment]
Fixed already in 2.0
